src/osp/Shaders/Flat.h

```diff
@@ -69,75 +68,60 @@ void draw_ent_flat(
-        active::TexGlEntStorage_t const&                        diffuse,
-        ACtxDrawFlat &rData)
+struct ArgsForSyncDrawEntFlat
 {
-    using namespace active;
+    active::DrawEntSet_t const&             hasMaterial;
+    active::RenderGroup::Storage_t *const   pStorageOpaque;
+    active::RenderGroup::Storage_t *const   pStorageTransparent;
+    active::DrawEntSet_t const&             opaque;
+    active::DrawEntSet_t const&             transparent;
+    active::TexGlEntStorage_t const&        diffuse;
+    ACtxDrawFlat&                           rData;
+};
+
+inline void sync_drawent_flat(active::DrawEnt ent, ArgsForSyncDrawEntFlat const args)
+{
+    using namespace osp::active;
+
+    auto const entInt = std::size_t(ent);
+
+    bool const hasMaterial = args.hasMaterial.test(entInt);
+    bool const hasTexture = (args.diffuse.size() > std::size_t(ent)) && (args.diffuse[ent].m_glId != lgrn::id_null<TexGlId>());
+
+    FlatGL3D *pShader = hasTexture
+                      ? &args.rData.shaderDiffuse
+                      : &args.rData.shaderUntextured;
 
-    for (; dirtyIt != dirtyLast; std::advance(dirtyIt, 1))
+    if (args.pStorageTransparent != nullptr)
     {
-        DrawEnt const ent = *dirtyIt;
-
-        // Erase from group if they exist
-        if (pStorageOpaque != nullptr)
-        {
-            pStorageOpaque->remove(ent);
-        }
-        if (pStorageTransparent != nullptr)
-        {
-            pStorageTransparent->remove(ent);
-        }
-
-        if ( ! hasMaterial.test(std::size_t(ent)))
-        {
-            continue; // Phong material is not assigned to this entity
-        }
-
-        Flat *pShader = (diffuse[ent].m_glId != lgrn::id_null<TexGlId>())
-                      ? &rData.m_shaderDiffuse
-                      : &rData.m_shaderUntextured;
-
-        if (drawBasic[ent].m_opaque)
-        {
-            if (pStorageOpaque == nullptr)
-            {
-                continue;
-            }
-
-            pStorageOpaque->emplace(ent, EntityToDraw{&draw_ent_flat, {&rData, pShader} });
-        }
-        else
-        {
-            if (pStorageTransparent == nullptr)
-            {
-                continue;
-            }
-
-            pStorageTransparent->emplace(ent, EntityToDraw{&draw_ent_flat, {&rData, pShader} });
-        }
+        auto value = (hasMaterial && args.transparent.test(entInt))
+                   ? std::make_optional(EntityToDraw{&draw_ent_flat, {&args.rData, pShader}})
+                   : std::nullopt;
+
+        storage_assign(*args.pStorageTransparent, ent, std::move(value));
+    }
+
+    if (args.pStorageOpaque != nullptr)
+    {
+        auto value = (hasMaterial && args.opaque.test(entInt))
+                   ? std::make_optional(EntityToDraw{&draw_ent_flat, {&args.rData, pShader}})
+                   : std::nullopt;
+
+        storage_assign(*args.pStorageOpaque, ent, std::move(value));
     }
 }
 
+template<typename ITA_T, typename ITB_T>
```

Comment:
minor: since we're C++20, you could use the C++20 template thing `std::sentinal_for<ITA_T> ITB_T` over `typename`

Reply:
This looks rather noisy for and gets passed to for_each anyways.

I'd actually prefer to limit the use of C++20, especially ranges. Consider that there's relatively few C++ developers in the gamedev world, and even fewer of them know C++20. It's already difficult enough to get new developers on this project.